src/index.ts

```diff
@@ -146,48 +146,70 @@ function getNewtype(model: CaseClass): Reader<Ctx, gen.CustomTypeDeclaration> {
+        const dependencies = interfaceDecl.properties
+          .map(p => gen.printStatic(p.type))
+          .filter(p => !caseClass.typeParams.map(p => p.name).includes(p));
+        return gen.customTypeDeclaration(
+          caseClass.name,
+          `export interface ${caseClass.name}<${caseClass.typeParams.map(p => p.name)}> ${gen.printStatic(
+            interfaceDecl
+          )}`,
+          `export const ${caseClass.name} = <${staticParams}>(${runtimeParams}) => ${gen.printRuntime(interfaceDecl)}`,
+          dependencies
         );
+      } else {
+        return gen.typeDeclaration(caseClass.name, interfaceDecl, true, isReadonly);
       }
-      const caseClass = model as CaseClass;
-      return traverseReader(caseClass.members, getProperty).map(properties => {
-        const interfaceDecl = gen.typeCombinator(properties, model.name);
-        if (caseClass.typeParams && caseClass.typeParams.length > 0) {
-          const staticParams = caseClass.typeParams.map(p => `${p.name} extends t.Any`).join(', ');
-          const runtimeParams = caseClass.typeParams.map(p => `${p.name}: ${p.name}`).join(', ');
-          const dependencies = interfaceDecl.properties
-            .map(p => gen.printStatic(p.type))
-            .filter(p => !caseClass.typeParams.map(p => p.name).includes(p));
-          return gen.customTypeDeclaration(
-            model.name,
-            `export interface ${model.name}<${caseClass.typeParams.map(p => p.name)}> ${gen.printStatic(
-              interfaceDecl
-            )}`,
-            `export const ${model.name} = <${staticParams}>(${runtimeParams}) => ${gen.printRuntime(interfaceDecl)}`,
-            dependencies
-          );
-        } else {
-          return gen.typeDeclaration(model.name, gen.typeCombinator(properties, model.name), true, isReadonly);
-        }
-      });
     })
   );
 }
 
+function getEnumDeclaration(enumModel: Enum): Reader<Ctx, gen.TypeDeclaration | gen.CustomTypeDeclaration> {
+  return reader.of(
+    gen.typeDeclaration(
+      enumModel.name,
+      gen.keyofCombinator(enumModel.values.map(v => v.name), enumModel.name),
+      true,
+      false
+    )
+  );
+}
+
+function getTaggedUnionDeclaration(
+  taggedUnion: TaggedUnion
+): Reader<Ctx, gen.TypeDeclaration | gen.CustomTypeDeclaration> {
+  const getTaggedUnionValue = (v: TaggedUnionValue): Reader<Ctx, gen.TypeReference> =>
+    traverseReader(v.params, getProperty).map(properties =>
+      gen.typeCombinator(properties.concat(gen.property('_type', gen.literalCombinator(v.name))), v.name)
+    );
+  return traverseReader(taggedUnion.values, getTaggedUnionValue).map(values =>
+    gen.typeDeclaration(taggedUnion.name, gen.unionCombinator(values, taggedUnion.name), true, false)
+  );
+}
+
+function getDeclarations(models: Array<Model>): Reader<Ctx, Array<gen.TypeDeclaration | gen.CustomTypeDeclaration>> {
+  return traverseReader(models, model => {
+    switch (model._type) {
```

Comment:
Qui mi sono messo a usare il campo `_type` per fare il dispatch perché il check come fatto prima (check di quali campi sono definiti su `model`) non andava bene per le tagged union (hanno gli stessi campi delle enum).

Mi sembra anche più semplice rispetto ai check sui campi definiti in `model` per cui l'ho usato in generale.

Ci sono controindicazioni? (a me sembra che metarpheus lo generi sempre e quindi non sia un problema usarlo)

Reply:
Mi sembra molto meglio così 👍

Reply:
Molto meglio così, anzi, sarebbe ripensare il modell di metarpheus da capo in modo da usare sempre tagged unions per i vari tipi di nodi, e a cascata fare un bel refactor di metarpheus-io-ts per usarlo.